spring-pulsar/src/main/java/org/springframework/pulsar/core/PulsarTemplate.java

```diff
@@ -115,22 +176,49 @@ private CompletableFuture<MessageId> doSendAsync(String topic, T message,
 			ProducerBuilderCustomizer<T> producerCustomizer) throws PulsarClientException {
 		final String topicName = ProducerUtils.resolveTopicName(topic, this.producerFactory);
 		this.logger.trace(() -> String.format("Sending msg to '%s' topic", topicName));
-		final Producer<T> producer = prepareProducerForSend(topic, message, messageRouter, producerCustomizer);
-		TypedMessageBuilder<T> messageBuilder = producer.newMessage().value(message);
-		if (typedMessageBuilderCustomizer != null) {
-			typedMessageBuilderCustomizer.customize(messageBuilder);
-		}
-		return messageBuilder.sendAsync().whenComplete((msgId, ex) -> {
-			if (ex == null) {
-				this.logger.trace(() -> String.format("Sent msg to '%s' topic", topicName));
-				// TODO success metrics
-			}
-			else {
-				this.logger.error(ex, () -> String.format("Failed to send msg to '%s' topic", topicName));
-				// TODO fail metrics
+
+		PulsarMessageSenderContext senderContext = PulsarMessageSenderContext.newContext(topicName, this.beanName);
```

Comment:
You need to think about moving this object creation already after `Observation.NOOP`.
See `createNotStarted()` which accept `Supplier<T> contextSupplier` and does `NOOP` check.
You might not need that logic in your code afterwards.

Reply:
Ah this would be nice. Thx. I will look into that here and below as well.

Reply:
Unfortunately ...

1. Those built-in checks in `createNotStarted` will not take into account my local concerns for `containerProperties.isObservationEnabled `nor the null OR. In fact, the null OR will call the method to NPE as it does not tolerate a null OR being passed in.

2. The ObservationDocumentation.observe that takes a supplier only checks if registry.isNoop() before calling get() on the supplier which it then passes into Observeration.createNotStarted and does the other checks as:
```
if (registry == null || registry.isNoop()
        || !registry.observationConfig().isObservationEnabled(observationConvention.getName(), context)
        || observationConvention == NoopObservationConvention.INSTANCE) {
    return NOOP;
} else ...
```

For this to be useful in my case, I would need to be able to pass in some predicate to determine if it should be noop and also allow the supplier to make it further in the call chain before calling get().

Reply:
Sorry, I don't understand why you talk about `containerProperties.isObservationEnabled`.
You use `Observation.NOOP` condition in your own code. My point is to move that logic out of your code and supply context in lazy manner.
I believe that we step into this observation logic here only if that `containerProperties.isObservationEnabled == true`.
So, why do you try to add this variable into an equation where I just talked about deferring context creation down to the `createNotStarted()` call?

Reply:
I understand what you are wanting to do here, get rid of this if/else and let the createNotStarted+supplier handle this for us. I too don't like this if/else/create-myself pattern that we have to do. 

However, here is why the current impl of createNotStarted+supplier does not really help our cause. 

We have 2 custom conditions that dictate if we are using Observations or not, a boolean `isObservationEnabled` property and whether or not there is a OR in the AC. We have to respect these properties. None of the observation API calls will allow a null ObsReg either. 

Lets look at the suggested code path: 
```java
default <T extends Observation.Context> Observation observation(@Nullable ObservationConvention<T> customConvention,
        @NonNull ObservationConvention<T> defaultConvention, @NonNull Supplier<T> contextSupplier,
        @NonNull ObservationRegistry registry) {
    if (registry.isNoop()) {
        return Observation.NOOP;
    }
    return observation(customConvention, defaultConvention, contextSupplier.get(), registry);
}
```
From https://github.com/micrometer-metrics/micrometer/blob/13fbc3e6d6d8359920f4e99ab3408067f6e7b68f/micrometer-observation/src/main/java/io/micrometer/observation/docs/ObservationDocumentation.java#L209-L216

This is the API. you are suggesting to use, correct? It is the only one I see that accepts supplier. 


As you can see it only checks `registry.isNoop()` (it would NPE here already if we don't check the OR for null) and then it calls supplier.get() before sending to observation() method that in turn calls into Observation.createNotStarted which finally (after a few method chains) does the other checks such as: 

```java
static <T extends Context> Observation createNotStarted(ObservationConvention<T> observationConvention,
        @Nullable T context, @Nullable ObservationRegistry registry) {
    if (registry == null || registry.isNoop()
            || !registry.observationConfig().isObservationEnabled(observationConvention.getName(), context)
            || observationConvention == NoopObservationConvention.INSTANCE) {
        return NOOP;
    }
  return new SimpleObservation(observationConvention, registry, context == null ? new Context() : context);
}
```
From 
https://github.com/micrometer-metrics/micrometer/blob/13fbc3e6d6d8359920f4e99ab3408067f6e7b68f/micrometer-observation/src/main/java/io/micrometer/observation/Observation.java#L222-L230

 
1. I do not see a way to get rid of our `if/else` as we need to do our custom checks
2. The supplier.get is called immediately so I don't think it is a win either

Maybe I am missing something?

Reply:
You do check for `OR` in your code, but you do not for `NOOP` and wrap your context object into `Supplier`.
That's how I see it for now.
Yes, Observation API might be improved with your investigation, but that's already outside of your `containerProperties.isObservationEnabled` condition.

Reply:
Got it. Yeh as it is we would not be able to get rid of the `if/else` and the `supplier` would buy little value as it is immediately invoked after only checking `registry.isNoop()`. It would need to be delayed until the actual `createNotStarted` to use that approach. 

I think I will leave it as-is based on my findings of the current API impl - unless you are strongly opposed :)